server/controllers/additional-sheet-parser.js

```diff
@@ -0,0 +1,91 @@
+// Parses "optimization_paramters," expression data sheets, and 2-column sheets
+// from GRNmap input or output workbook
+
+var TWO_COL_SHEET_NAMES = [
+    "production_rates",
+    "degradation_rates",
+    "threshold_b",
+    "optimized_production_rates",
+    "optimized_threshold_b"];
+
+var EXPRESSION_SHEET_SUFFIXES = ["_expression", "_optimized_expression", "_sigmas"];
+
+var isExpressionSheet = function (sheetName) {
+    return EXPRESSION_SHEET_SUFFIXES.some(function (suffix) {
+        return sheetName.includes(suffix);
+    });
+};
+
+var fillArray = function (value, array, length) { // mutator
+    while (array.length < length) {
+        array.push(value);
+    }
+    return array;
+};
+
+var parseMetaDataSheet = function (sheet) {
+    var meta = {};
+    sheet.data.forEach(function (element, index) {
+        if (index !== 0) {
+            var value = element.slice(1);
+            // Extract element from array if array contains only 1 value
+            meta[element[0]] = value.length > 1 ? value : value[0];
+        }
+    });
+    return meta;
+};
+
+var parseTwoColumnSheet = function (sheet) {
+    var data = {};
+    sheet.data.forEach(function (element, index) {
+        if (index !== 0) {
+            data[element[0]] = element[1];
+        }
+    });
+    return data;
+};
+
+var parseExpressionSheet = function (sheet) {
+    var expressionData = {};
+    var geneData = {};
+    expressionData["time_points"] = sheet.data[0].slice(1);
+    var numberOfDataPoints = expressionData["time_points"].length;
+    for (var j = 1; j < sheet.data.length; j++) {
+        var geneName = sheet.data[j][0];
+        if (geneName) {
+            var rowData = sheet.data[j].slice(1);
+            // Sometimes, missing data is at the end of the row. In this case, pad the
+            // array with nulls
+            if (rowData.length < numberOfDataPoints) {
+                fillArray(null, rowData, numberOfDataPoints);
+            }
+            geneData[geneName] = rowData;
+        }
+    }
+    expressionData["data"] = geneData;
+    return expressionData;
+};
+
+module.exports = function (workbook) {
+
+    var output = {
+        expression: {}, // expression data
+        meta: {},
+        test: {} // 2-column data
+    };
+
+    for (var i = 0; i < workbook.length; i++) {
+        var sheet = workbook[i];
```

Comment:
This is a classic pattern that, in more modern languages, is done as an iterator function rather than a `for` loop: index variable (`i` here) is only used in the beginning, and the value it represents is assigned right away (`sheet`), with that variable being the only one used for the rest of the loop.

When you spot that, it’s more compact (and some might say direct) to write:
```javascript
workbook.forEach(function (sheet) {
    // .....etc.....
})
```